### ISSUE TITLE
GCC 4.8/4.9, -std=c++11 and "error: expected ‘)’ before __builtin_ia32_pclmulqdq128"

### ISSUE THREAD

**Issue by noloader** [weidai11/cryptopp#206]
We have a couple of issues under GCC 4.8 and 4.9 when using `-std=c++11` and `-std=c++14`. The issue is the instructions `pclmulqdq128` and `aeskeygenassist128` used in inline assembly cannot compile:

```
$ make CXXFLAGS="-DDEBUG -g3 -O0 -std=c++11 -fPIC -march=native -pipe" vmac.o
g++ -DDEBUG -g3 -O0 -std=c++11 -fPIC -march=native -pipe -c vmac.cpp
In file included from /usr/lib/gcc/x86_64-linux-gnu/4.9/include/x86intrin.h:43:0,
                 from /usr/include/x86_64-linux-gnu/c++/4.9/bits/opt_random.h:33,
                 from /usr/include/c++/4.9/random:50,
                 from /usr/include/c++/4.9/bits/stl_algo.h:66,
                 from /usr/include/c++/4.9/algorithm:62,
                 from stdcpp.h:13,
                 from cryptlib.h:87,
                 from vmac.h:9,
                 from vmac.cpp:7:
cpu.h:96:1: error: expected ‘)’ before ‘__builtin_ia32_pclmulqdq128’
 _mm_clmulepi64_si128 (__m128i a, __m128i b, const int i)
 ^
cpu.h:102:1: error: expected ‘)’ before ‘__builtin_ia32_aeskeygenassist128’
 _mm_aeskeygenassist_si128 (__m128i a, const int i)
 ^
GNUmakefile:735: recipe for target 'vmac.o' failed
make: *** [vmac.o] Error 1
```

The unexpected ‘)’ is due to an intermediate cast to `__v2di`, which appears to be missing under Debug builds (`-O0`) when using C++11 (C++03 is OK):

```
$ grep -IR '__builtin_ia32_pclmulqdq128' /usr/lib
/usr/lib/gcc/x86_64-linux-gnu/4.9/include/wmmintrin.h:  return (__m128i) __builtin_ia32_pclmulqdq128 ((__v2di)__X,
/usr/lib/gcc/x86_64-linux-gnu/4.9/include/wmmintrin.h:  ((__m128i) __builtin_ia32_pclmulqdq128 ((__v2di)(__m128i)(X), 
```

There's an open question on Stack Overflow at [Where is __v2di declared when using -std=c++11 under GCC?](http://stackoverflow.com/q/38051959)

**Comment by noloader** [comment]
Cleared at [Commit 6c88c3bf6526c857](http://github.com/weidai11/cryptopp/commit/6c88c3bf6526c857dbae2c23febac00081571a48).

**Comment by noloader** [comment]
Reopened because the fix broke GCC 4.8 and 4.9.

**Comment by noloader** [comment]
This is fixed in the [cpu dev-branch](http://github.com/weidai11/cryptopp/tree/cpu). The problem is, Clang is broken for x86_64 in the branch. Everything else is working fine, including GCC, ICPC, VC++ on i686, x86_64, ARM32, ARM64 and MIPS.

**Comment by noloader** [comment]
Testing with `-std=c++14` shows the same issue:

```
g++ -DDEBUG -g3 -O0 -std=c++14   -Wno-deprecated-declarations -march=native -pipe -D_GLIBCXX_DEBUG -c blake2.cpp
In file included from /usr/lib/gcc/i586-linux-gnu/4.9/include/x86intrin.h:41:0,
                 from /usr/include/i386-linux-gnu/c++/4.9/bits/opt_random.h:33,
                 from /usr/include/c++/4.9/random:50,
                 from /usr/include/c++/4.9/bits/stl_algo.h:66,
                 from /usr/include/c++/4.9/algorithm:62,
                 from stdcpp.h:13,
                 from cryptlib.h:87,
                 from blake2.cpp:8:
cpu.h:96:1: error: expected unqualified-id before ‘int’
 _mm_extract_epi32 (__m128i a, const int i)
 ^
cpu.h:96:1: error: expected ‘)’ before ‘int’
cpu.h:96:1: error: expected ‘)’ before ‘int’
cpu.h:103:1: error: expected ‘)’ before ‘__builtin_ia32_vec_set_v4si’
 _mm_insert_epi32 (__m128i a, int b, const int i)
 ^
In file included from /usr/lib/gcc/i586-linux-gnu/4.9/include/x86intrin.h:43:0,
                 from /usr/include/i386-linux-gnu/c++/4.9/bits/opt_random.h:33,
                 from /usr/include/c++/4.9/random:50,
                 from /usr/include/c++/4.9/bits/stl_algo.h:66,
                 from /usr/include/c++/4.9/algorithm:62,
                 from stdcpp.h:13,
                 from cryptlib.h:87,
                 from blake2.cpp:8:
cpu.h:117:1: error: expected ‘)’ before ‘__builtin_ia32_pclmulqdq128’
 _mm_clmulepi64_si128 (__m128i a, __m128i b, const int i)
 ^
cpu.h:123:1: error: expected ‘)’ before ‘__builtin_ia32_aeskeygenassist128’
 _mm_aeskeygenassist_si128 (__m128i a, const int i)
 ^
```

**Comment by noloader** [comment]
Here's part of the answer from the [GCC manual, 3.7.2 Common Predefined Macros](http://gcc.gnu.org/onlinedocs/cpp/Common-Predefined-Macros.html):

> `__OPTIMIZE__`
> `__OPTIMIZE_SIZE__`
> `__NO_INLINE__`
> 
> These macros describe the compilation mode. `__OPTIMIZE__` is defined in all optimizing compilations. `__OPTIMIZE_SIZE__` is defined if the compiler is optimizing for size, not speed. `__NO_INLINE__` is defined if no functions will be inlined into their callers (when not optimizing, or when inlining has been specifically disabled by -fno-inline).
> 
> These macros cause certain GNU header files to provide optimized definitions, using macros or inline functions, of system library functions. You should not use these macros in any way unless you make sure that programs will execute with the same effect whether or not they are defined. If they are defined, their value is 1.

**Comment by noloader** [comment]
I was never able to track down the bad interaction. I tried to diff the preprocessed source files (`-E` instead of `-c`), but there were thousands of lines of differences between a "good" and a "bad".

I talked to Wei about the issue. The functionality the library was providing through its own definition of the instrinsics was of marginal value in 2016. In contrast, they were of great value in 2006 or 2008, before all compilers provided them. The compilers being used back then were GCC 4.0.3 to about GCC 4.3.

Cleared at [Commit fb6a11ff08b9277e, Remove library supplied aesenc, aesdec and friends (Issue 206)](http://github.com/weidai11/cryptopp/commit/fb6a11ff08b9277e9f7a2fecae83027d3e0e60df).
